ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,529 @@
+	if len(tries) == 0 {
+		return 0
+	}
+	estimatedTrieNodeCount := 2*int(tries[0].AllocatedRegCount()) - 1
+	return estimatedTrieNodeCount / subtrieCount
+}
+
+func storeSubTrieConcurrently(
+	subtrieRoots [subtrieCount][]*node.Node,
+	estimatedSubtrieNodeCount int,
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+) (
+	map[*node.Node]uint64, // node indices
+	uint64, // node count
+	[]uint32, //checksums
+	error, // any exception
+) {
+	logger.Info().Msgf("storing %v subtrie groups with average node count %v for each subtrie", subtrieCount, estimatedSubtrieNodeCount)
+
+	resultChs := make([]chan *resultStoringSubTrie, 0, len(subtrieRoots))
+	for i, subTrieRoot := range subtrieRoots {
+		resultCh := make(chan *resultStoringSubTrie)
+		go func(i int, subTrieRoot []*node.Node) {
+			roots, nodeCount, checksum, err := storeCheckpointSubTrie(
+				i, subTrieRoot, estimatedSubtrieNodeCount, outputDir, outputFile, logger)
+			resultCh <- &resultStoringSubTrie{
+				Index:     i,
+				Roots:     roots,
+				NodeCount: nodeCount,
+				Checksum:  checksum,
+				Err:       err,
+			}
+			close(resultCh)
+		}(i, subTrieRoot)
+		resultChs = append(resultChs, resultCh)
+	}
+
+	logger.Info().Msgf("subtrie roots have been stored")
+
+	results := make(map[*node.Node]uint64, 1<<(subtrieLevel+1))
+	results[nil] = 0
+	nodeCounter := uint64(0)
+	checksums := make([]uint32, 0, len(results))
+	for _, resultCh := range resultChs {
+		result := <-resultCh
+		if result.Err != nil {
+			return nil, 0, nil, fmt.Errorf("fail to store %v-th subtrie, trie: %w", result.Index, result.Err)
+		}
+
+		for root, index := range result.Roots {
+			// the original index is relative to the subtrie file itself.
+			// but we need a global index to be referenced by top level trie,
+			// therefore we need to add the nodeCounter
+			results[root] = index + nodeCounter
+		}
+		nodeCounter += result.NodeCount
+		checksums = append(checksums, result.Checksum)
+	}
+
+	return results, nodeCounter, checksums, nil
+}
+
+func createWriterForTopTries(dir string, file string, logger *zerolog.Logger) (io.WriteCloser, error) {
+	fullPath, topTriesFileName := filePathTopTries(dir, file)
+	if utilsio.FileExists(fullPath) {
+		return nil, fmt.Errorf("checkpoint file for top tries %s already exists", fullPath)
+	}
+
+	return createClosableWriter(dir, logger, topTriesFileName, fullPath)
+}
+
+func createWriterForSubtrie(dir string, file string, logger *zerolog.Logger, index int) (io.WriteCloser, error) {
+	fullPath, subTriesFileName, err := filePathSubTries(dir, file, index)
+	if err != nil {
+		return nil, err
+	}
+	if utilsio.FileExists(fullPath) {
+		return nil, fmt.Errorf("checkpoint file for %v-th sub trie %s already exists", index, fullPath)
+	}
+
+	return createClosableWriter(dir, logger, subTriesFileName, fullPath)
+}
+
+func createClosableWriter(dir string, logger *zerolog.Logger, fileName string, fullPath string) (io.WriteCloser, error) {
```

Comment:
nit: you likely don't need to thread dir through all function calls here: you can just use `filepath.Dir()` here.  File moves are guaranteed to be atomic only within a directory anyway.